### DISCUSSION TITLE
Cloud Hosting (Longterm deployment strategy)

### DISCUSSION THREAD

**Discussion by sacdallago** [PathwayCommons/factoid#451] [General]
### initial list

- [x] Digital Ocean
- [x] Heroku
- [x] A2 Hosting
- [x] Linode
- [ ] Azure @sacdallago 

#### Factors to consider
- Cost
- Level of support
- Ongoing maintenance costs
- Level of service (uptime, reliability)

Comment on this issue explaining how the services address the factors above.

**Comment by jvwong** [comment]
## [Digital Ocean](https://www.digitalocean.com/)

### [Pricing](https://www.digitalocean.com/pricing/)

#### Basic Droplets

| Memory | vCPUs | *Transfer | SSD Disk | $/HR | $/MO |
| -- | -- | -- | -- | -- | -- |
| 4GB | 2vCPUs | 4TB | 80GB | $0.030 | $20 |
| 8GB | 4vCPUs | 5TB | 160GB | $0.060 | $40 |
> * Droplets include free outbound data transfer, starting at 1,000 GiB/month for the smallest plan....Transfer allowance is pooled cumulatively at the account level, not individually at the Droplet level.

#### General Purpose Droplets

| Memory | vCPUs | *Transfer | SSD Disk | $/HR | $/MO |
| -- | -- | -- | -- | -- | -- |
| 8GB | 2vCPU | 4TB | 25GB | 0.089 | $60 | 
|  16GB | 4vCPUs | 5TB | 50GB | $0.179 | $120 | 

### [Service Level Agreement](https://www.digitalocean.com/docs/platform/droplet-policies/)

> DigitalOcean provides a 99.99% uptime SLA for both Droplets and block storage. We refund lost time back to your account at the hourly rate incurred.

### [Support](https://www.digitalocean.com/products/support/)

> 24/7 Support 
> Have a question? Send us a note using the form below and someone from the DigitalOcean team will be in touch soon.

### Usability
> [Sign up now with a $100 credit.](https://try.digitalocean.com/frontend/)

( Probably need to try it out to see what's what. )


---


## [Linode](https://www.linode.com/)

### [Pricing](https://www.linode.com/pricing/)

#### Shared Plans 

| RAM | CPU | Storage | Transfer | Network In | Network Out | Price |
| -- | -- | -- | -- | -- | -- | -- |
| 4 GB | 2 Cores | 80 GB SSD | 4 TB | 40 Gbps | 4000 Mbps | $20 / mo |
| 8GB | 4 Cores | 160 GB SSD | 5 TB | 40 Gbps | 5000 Mbps | $40 / mo |

#### Dedicated CPU Plans 
| RAM | CPU | Storage | Transfer | Network In | Network Out | Price |
| -- | -- | -- | -- | -- | -- | -- |
| 4 GB | 2 Cores | 80 GB SSD | 4 TB | 40 Gbps | 4000 Mbps | $30 / mo | 
| 8GB |  4 Cores | 160 GB SSD | 5 TB | 40 Gbps | 5000 Mbps | $60 / mo | 

### [Service Level Agreement](https://www.linode.com/legal-sla/)

> Uptime Guarantee. We guarantee 99.9% monthly Covered Services uptime
> Subject to the Terms of Service and this SLA, the Account Holder may request a pro-rata credit for any affected monthly billing period in which we are unable to satisfy the Uptime Guarantee (collectively “Service Credits”). 

### [Support](https://www.linode.com/support/)

> Call Support 855-4-LINODE 
> Chat on IRC 
> Contact Linode Sales  844-869-6072
...

### Usability

( Probably need to try it out to see what's what. )


---


## [Heroku](https://www.heroku.com/)

### [Pricing](https://www.heroku.com/pricing)

> The containers used at Heroku are called “dynos.” Dynos are isolated, virtualized Linux containers that are designed to execute code based on a user-specified command. 

| Dyno Type | Memory (RAM) | CPU Share | Compute | Dedicated | Sleeps |
| -- | -- | -- | -- | -- | -- |
| hobby | 512 MB | 1x | 1x-4x | no | no |
| standard-1x | 512 MB | 1x | 1x-4x | no | no|
| standard-2x | 1024 MB | 2x | 4x-8x | no | no |
| performance-m | 2.5 GB | 100% | 12x | yes | no |
| performance-l | 14 GB | 100% | 50x | yes | no | 

| Dyno Type | Price per dyno/month
| -- | -- | 
| free | $0 | 
| hobby | $7 | 
| standard-1x | $25 | 
| standard-2x | $50|  
| performance-m | $250 | 
| performance-l | $500 | 

### [Service Level Agreement](https://www.linode.com/legal-sla/)

### [Support](https://www.heroku.com/support#plans)

> Standard
> Hours: 12x5
> Select a topic, then pick a category and we'll show you commonly asked questions and answers.
Looking for more help? Ask the community or create a ticket to get it routed to the best person to answer it. 

### Usability
( Probably need to try it out to see what's what. )


---


## [A2 Hosting](https://www.a2hosting.ca/vps-hosting/)

### [Pricing](https://www.a2hosting.ca/vps-hosting/unmanaged/compare)

## Unmanaged VPS 

| Plan | Runway 1 | Runway 2 | Runway 4 | Supersonic 8 |
| -- | -- | -- | -- | -- |
| RAM | 1 GB  | 2GB  | 4 GB  |8 GB  |
|  RAID-10 SSD Storage | 150 GB SSD  |250 GB SSD  | 450 GB SSD  | 150 GB NVMe  |
|  Transfer | 2 TB  |  3 TB  | 4 TB  | 2 TB  |
|  Cores | 1 | 2 | 4 | 2 |
| Reneal Cost /mo |  $11.47 |  $19.12 |  $25.20 |  $76.53 | 

### [Service Level Agreement](https://www.a2hosting.com/99.9-uptime-commitment)

> Our 24/7/365 support staff with expert system admins keep your server running at its peak. We use industry best practices, the best data centers and ensure your server has the latest security patches. That's why we can commit to your site being up over 99.9% of the time.

### [Support](https://my.a2hosting.com/submitticket.php)

> Submit a Support Ticket

### Usability
( Probably need to try it out to see what's what. )

**Reply by maxkfranz** [reply]
Re. usability, it looks like we'd have a setup similar to today.  A VM would have to be manually managed using docker compose for deployment.

One alternative is to include the app deployment offered by Digital Ocean, where they deploy a docker per Github repo.  That requires one VM per docker image.  For Biofactoid, that would multiply the monthly cost (factoid 1x, grounding 2x, convertors 3x, rethinkdb 4x, ...).  In this case, each app gets automatically deployed when the master branch is updated on Github.

**Comment by jvwong** [comment]
I know we had Digital Ocean on our list anyways to try out, this just serves as due diligence.

**Reply by maxkfranz** [reply]
I tried out DO on Friday.  The simplest way to do it there would be to just spin up a single VM for Biofactoid prod.  We'd start it up manually using docker compose, as we do on the existing cluster.  We could make it easier to maintain the DO VM by setting a @reboot cron job so that the dockers spin up when the VM is rebooted.  That way, anyone can just hit the restart button in the DO panel if there's a problem.

**Reply by jvwong** [reply]
What do you think we need in terms of specs to run our app?

**Reply by maxkfranz** [reply]
I tried the basic one, and it was pretty slow because of the grounding indexing.  It was swapping all the time.  We probably need a higher option with more RAM for Biofactoid.  I suspect we need at least 8GB, maybe 16GB.

**Reply by jvwong** [reply]
This indexing thing is the biggest pain. Before the rethinkdb dump/restore scripts, I had this thing that dumped and restored the elasticsearch index. I want to revive that if possible - that way, we could just store the data (zenodo) and pop it in...This would reduce our requirements too.

**Reply by maxkfranz** [reply]
It would be good to have prebuilt indices in any case, but it might not reduce the requirements too much.  For something like the grounding search to be fast, you want to keep all the DB in memory if possible.  The data is >10GB, so we may end up needing the 16GB option anyway.  I expect the minimum would be 8GB even with prebuilt indices.

**Comment by sacdallago** [comment]
I have had really good experience with them: https://www.hosteurope.de/en/Server/Virtual-Server (the 16GB/RAM version is 40E a month, and you got root -- it doesn't "scale" as easy as AWS/Azure/..., but you can still upgrade and round robin ;))

Slightly different, but consider reading through https://github.com/zszazi/Deep-learning-in-cloud?s=09#deploy-your-model-as-a-web-app and subsequent sections.

**Reply by sacdallago** [reply]
Ohter plus w.r.t. other things I've seen around:
- No traffic cap
- decent # of cores (although virtual)
- can create backups

Only problem for you guys: hosting in Germany (aka, some latency for Canada/US)

**Comment by jvwong** [comment]
Testing observations for Digital Ocean. From sign up to running biofactoid via Docker Compose, took less than 45 minutes, which includes setting up users, ssh keys, installing Docker/Compose. This doesn't  include restoring data or grounding-search data, however. Really good docs.

### Droplet
8 GB Memory / 160 GB Disk / TOR1 - Ubuntu 18.04 (LTS) x64

### Indexing: 

| Source        |  Time to update (min) |
| ----------- | :-----------------: |
| ChEBI          |  1 |
| NCBI Gene  |  1 |
| UniProt        |  6 |

**Comment by jvwong** [comment]
Additional considerations:

- Proxy (nginx)
- Data backups
- Certificates/renewal

**Comment by jvwong** [comment]
'To the cloud' (Digital Ocean) task list:

- [x] Provision DO Droplet (8 GB Memory / 160 GB Disk / TOR1 - Ubuntu 18.04 (LTS) x64)
  - Droplet owned by team 'biofactoid'
- [x] Run Dockerized instance 
- [x] Self-restart
  - [x] proxy
  - [x] webapp
  - [x] index
  - [x] grounding
  - [x] converter
  - [x] db
- [x] Certificates: Certbot/Let's Encrypt
- [x] [GitHub actions](https://github.com/features/actions) and the [DO CLI](https://www.digitalocean.com/community/tools/digitalocean-command-line-interface) for semi-automated deployment
- [x] Database backup
  - Approach? https://github.com/PathwayCommons/rethinkdb-backup/issues/2
- [x] Point master.factoid.baderlab.org at DO Droplet
- [ ] Point biofactoid.org at DO Droplet 
- [x] SSH to automate deployment to the droplet

**Comment by jvwong** [comment]
Docker stats for master.factoid.baderlab.org instance on Digital Ocean:

|  NAME  | CPU   %  | MEM  USAGE / LIMIT | MEM  % |
|--------------------|-----|--------|---|
|index               |0.53%|2.739GiB / 7.775GiB|35.23%|
|grounding           |0.03%|156.2MiB /  7.775GiB|1.96% |
|watch               |0.00%|6.832MiB /  7.775GiB|0.09% |
|proxy               |0.00%|5.793MiB /  7.775GiB|0.07% |
|webapp              |1.40%|331MiB  /  7.775GiB|4.16% |
|kibana              |0.22%|278.5MiB /  7.775GiB|3.50% |
|dbbackup            |0.02%|137.8MiB /  7.775GiB|1.73% |
heartbeat           |0.03%|12.96MiB /  7.775GiB|0.16% |
|converters_container|0.09%|179.9MiB /  7.775GiB|2.26% |
|sync                |0.01%|83.47MiB /  7.775GiB|1.05% |
|db                  |0.05%|424.3MiB /  7.775GiB|5.33% |


Disk:

| Filesystem    |  Size |  Used | Avail | Use%  |
|------------|--------------------|-----|--------|---|
| /dev/vda1     |   167G |    40G |  127G |   24%  |

RAM:

|                    | total |  used |  free |  shared  | buff/cache  | available |
|------------|------|-----|--------|------------|------|-----|
|Mem: |           7961 |        4606    |     729   | 1 |        2625 |       2968 |


> Here’s what each column means:
    total - This number represents the total amount of memory that can be used by the applications.
    used - Used memory. It is calculated as: used = total - free - buffers - cache
    free - Free / Unused memory.
    shared - This column can be ignored as it has no meaning. It is here only for backward compatibility.
    buff/cache - The combined memory used by the kernel buffers and page cache and slabs. This memory can be reclaimed at any time if needed by the applications. If you want buffers and cache to be displayed in two separate columns, use the -w option.
    available - An estimate of the amount of memory that is available for starting new applications, without swapping

**Comment by jvwong** [comment]
Summary of instances:

| Name                 |  Provider                 | Domain(s)                                   | DockerHub tag  |  Check for updates    |
|-----------------|--------------------|--------------------------------|-----------------|----------------------|
| development      | Donnelly                 | unstable.factoid.baderlab.org   |   latest                |  every 15 min             |
|                            |                                 | grounding.baderlab.org            |                             |                                   |
|                            |                                 | biopax.baderlab.org/convert    |                            |                                    |
| staging               | Digital Ocean         | master.factoid.baderlab.org      |   production        |  -                       |
| production         | Digital Ocean (TODO)         | biofactoid.org                             |   production        |  -                 |

**Comment by jvwong** [comment]
## Plan for moving biofactoid.org from 'Donnelly' to 'Digital Ocean'

Currently, on Digital Ocean (DO) we have: 
  - floating IP 174.138.112.180 intended for Droplet `biofactoid-staging`
  - floating IP 159.203.55.214 tied to a Droplet `biofactoid-production`

Below, we describe a process to test a full biofactoid install on `biofactoid-production` and migrate the (www.)biofactoid.org domain to `biofactoid-production`. 

- Install and access `biofactoid-production` via test domain
  - [x] Create a [factoid-docker-config](https://github.com/PathwayCommons/factoid-docker-config) branch  (i.e. 'beta') that supports the domain 'beta.biofactoid.org'
  - [x] Create test domain 'beta.biofactoid.org' and point at floating IP (159.203.55.214) linked to `biofactoid-production` 
  - [x] Trigger GitHub install workflow (beta branch) for 'beta.biofactoid.org' on `biofactoid-production` (DO)
  - [x] Restore the latest RethinkDB dump on `biofactoid-production` (DO)
    - [x] Backup to Dropbox `/backup/beta` OK?
    - [x] Restore local DB from backup OK? 
- Migrate biofactoid.org to `biofactoid-production`
  - [x] Serve a **maintenance page for (www.)biofactoid.org (Donnelly)
  - [x] Point (www.)biofactoid.org at floating IP (159.203.55.214) for `biofactoid-production` 
  - [x] Trigger GitHub install workflow (main branch) (DO)
  - [x] Restore the latest RethinkDB dump on `biofactoid-production` (DO)
- Cleanup 
  - [x] Delete the `biofactoid-staging` Droplet, but keep the floating IP (174.138.112.180) (DO)
  - [x] Point beta.biofactoid.org at floating IP 174.138.112.180, intended for `biofactoid-staging`, eventually 
    - [x] Disable/remove master.factoid.baderlab.org so it doesn't point at 174.138.112.180 (`biofactoid-staging`)
  - [x] Disable CRON jobs for RethinkDB dump (X.X.X.195) (Donnelly) 
  - [x] Update cerbot to remove migrated domains (X.X.X.153) (Donnelly) 
  - [x] Update internal DNS for relevant domains (www.)biofactoid.org (Donnelly) 

<hr/>

** Maintenance page
<img width="600" alt="Screenshot at Aug 20 09-31-24" src="https://user-images.githubusercontent.com/4706307/130241021-c44b86d2-804a-4908-a99b-38e28985f0f9.png">
